pennylane/circuit_drawer.py

```diff
@@ -0,0 +1,916 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the CircuitDrawer class which is used to draw CircuitGraph instances.
+"""
+import abc
+import math
+from collections import OrderedDict
+
+import numpy as np
+
+import pennylane as qml
+
+# pylint: disable=too-many-branches,too-many-arguments,too-many-return-statements,too-many-statements,consider-using-enumerate,too-many-instance-attributes
+
+
+def _transpose(target_list):
+    """Transpose the given list of lists.
+
+    Args:
+        target_list (list[list[object]]): List of list that will be transposed
+
+    Returns:
+        list[list[object]]: Transposed list of lists
+    """
+    return list(map(list, zip(*target_list)))
+
+
+class Grid:
+    """Helper class to manage Gates aligned in a grid.
+
+    The rows of the Grid are referred to as "wires",
+    whereas the columns of the Grid are reffered to as "layers".
+
+    Simultaneous access to both layers and wires via indexing is provided
+    by keeping both the raw grid (wires are indexed) and the transposed raw grid
+    (layers are indexed).
+
+    Args:
+        raw_grid (list, optional): Raw grid from which the Grid instance is built. Defaults to [].
+    """
+
+    def __init__(self, raw_grid=None):
+        if raw_grid is None:
+            raw_grid = []
+
+        self.raw_grid = raw_grid
+        self.raw_grid_transpose = _transpose(raw_grid)
```

Comment:
How about making this one a property?

```python
@property
def raw_grid_transpose(self):
    return _transpose(self.raw_grid)
```
This way it could also be avoided that each time the transposed ``raw_grid`` needs to be created.

Reply:
It is intentionally not a property so that I only do a transpose operation (which is costly) when something inside the grid changes